### PR TITLE
feat(sdk): output non nullable types

### PR DESCRIPTION
**Pull request by littlemooon**
Update SDK output to respect non-nullable graphql fields.

### CODE REVIEW

File: packages/codegen-doc/src/utils.ts [hunk 1]
```diff
@@ -38,6 +38,25 @@ export function reduceListType(
     : undefined;
 }
 
+/**
+ * Get the non null type name from any type node
+ */
+export function reduceNonNullType(
```
Comment (littlemooon) [inline packages/codegen-doc/src/utils.ts:44]: Find the non-nullable type if it exists

File: packages/codegen-sdk/src/model-visitor.ts [hunk 2]
```diff
@@ -95,6 +96,7 @@ export class ModelVisitor {
         const name = node.name.value;
         const type = printTypescriptType(this._context, node.type, Sdk.NAMESPACE);
         const query = findQuery(this._context, node);
+        const nonNull = Boolean(reduceNonNullType(node.type));
```
Comment (littlemooon) [inline packages/codegen-sdk/src/model-visitor.ts:99]: Add nonNull info to the model output

File: packages/codegen-sdk/src/parse-operation.ts [hunk 3]
```diff
@@ -84,6 +85,22 @@ export function parseOperations(
...
     const modelName = model?.name ?? "UNKNOWN_MODEL";
 
+    /** Find a parent operation */
+    const parentSdkKey = sdkPath.slice(0, -1).join("_");
+    const parent = acc[parentSdkKey]?.operations.find(operation => operation.path.join("_") === sdkKey);
+
+    /** Build the path through the response to the data */
+    const responsePath = parent
+      ? printList([parent?.print.responsePath, returnedField?.name?.value], parent?.nonNull ? `.` : "?.")
+      : printList([Sdk.RESPONSE_NAME, returnedField?.name?.value], ".");
```
Comment (littlemooon) [inline packages/codegen-sdk/src/parse-operation.ts:95]: Use the parent to produce a path through the response, to the data, without `?` where possible

File: packages/codegen-sdk/src/parse-operation.ts [hunk 4]
```diff
@@ -84,6 +85,22 @@ export function parseOperations(
...
+
+    /** Identify whether the operation response data is non-nullable */
+    const nonNullField = reduceNonNullType(
+      parent?.model?.fields.all.find(f => f.name === returnedField?.name?.value)?.node.type ?? ""
+    );
+    const nonNullQuery = query?.type && reduceNonNullType(query?.type);
+    const nonNull = Boolean(nonNullQuery || ((parent ? parent?.nonNull : true) && nonNullField));
```
Comment (littlemooon) [inline packages/codegen-sdk/src/parse-operation.ts:102]: We need to ensure either the operation is a query, that returns a non nullable. Or that the field from the parent is non nullable and the parent is also non nullable

File: packages/codegen-sdk/src/parse-operation.ts [hunk 5]
```diff
@@ -105,13 +122,11 @@ export function parseOperations(
       /** The name of the model in a list, if a list */
       list: listType?.replace("[]", ""),
       /** The returned promise result from fetch  */
-      promise: `${Sdk.FETCH_TYPE}<${listType ?? modelName}>`,
+      promise: `${Sdk.FETCH_TYPE}<${listType ?? modelName}${nonNull ? "" : " | undefined"}>`,
+      /** The typescript safe path through the response to the data */
+      responsePath,
```
Comment (littlemooon) [inline packages/codegen-sdk/src/parse-operation.ts:127]: Add response path for output later

File: packages/codegen-sdk/src/parse-operation.ts [hunk 6]
```diff
@@ -166,6 +181,7 @@ export function parseOperations(
       fetchArgs,
       parent,
       print,
+      nonNull,
```
Comment (littlemooon) [inline packages/codegen-sdk/src/parse-operation.ts:184]: Add nonNull info to operations for use later

File: packages/codegen-sdk/src/print-connection.ts [hunk 7]
```diff
@@ -31,13 +31,16 @@ export function isConnectionModel(model?: SdkModel): boolean {
...
 function printAbstractConnection(): string {
   return printLines([
-    printComment([
-      "Abstract class for connection models containing a list of nodes and pagination information",
-      "Follows the Relay spec",
-    ]),
-    `export abstract class ${Sdk.CONNECTION_TYPE}<${Sdk.NODE_TYPE}> extends ${Sdk.REQUEST_CLASS} {
-      public ${Sdk.PAGEINFO_NAME}?: ${Sdk.PAGEINFO_TYPE}
-      public ${Sdk.NODE_NAME}?: ${Sdk.NODE_TYPE}[]
+    printComment(["Connection models containing a list of nodes and pagination information", "Follows the Relay spec"]),
```
Comment (littlemooon) [inline packages/codegen-sdk/src/print-connection.ts:34]: This is no longer abstract as we need to implement a default PageInfo shape

File: packages/codegen-sdk/src/print-connection.ts [hunk 8]
```diff
@@ -58,15 +61,15 @@ function printConnectionVariables(): string {
 function printFetchType(): string {
   return printLines([
     printComment(["Fetch return type wrapped in a promise"]),
-    `export type ${Sdk.FETCH_TYPE}<${Sdk.RESPONSE_TYPE}> = Promise<${Sdk.RESPONSE_TYPE} | undefined>`,
+    `export type ${Sdk.FETCH_TYPE}<${Sdk.RESPONSE_TYPE}> = Promise<${Sdk.RESPONSE_TYPE}>`,
```
Comment (littlemooon) [inline packages/codegen-sdk/src/print-connection.ts:64]: The `LinearFetch` response type is now just a `Promise`, nothing else. This means we need to add `| undefined` to a few bits of the codegen where fields are not non-nullable.

File: packages/codegen-sdk/src/print-connection.ts [hunk 9]
```diff
@@ -79,13 +82,13 @@ export function printConnection(): string {
     {
       name: Sdk.NODE_NAME,
       type: `${Sdk.NODE_TYPE}[]`,
-      optional: true,
+      optional: false,
```
Comment (littlemooon) [inline packages/codegen-sdk/src/print-connection.ts:85]: Make `LinearConnection` use required `pageInfo` and `nodes`

File: packages/codegen-sdk/src/print-model.ts [hunk 10]
```diff
@@ -76,37 +77,42 @@ function printModel(context: SdkPluginContext, model: SdkModel): string {
...
-            model.fields.scalar.map(field => printSet(`this.${field.name}`, `${printModelScalar(field)} ?? undefined`))
+            model.fields.scalar.map(field =>
+              printSet(`this.${field.name}`, `${printModelScalar(field)}${field.nonNull ? "" : " ?? undefined"}`)
+            )
           ),
           printDebug("fields.scalarList"),
           printLines(
             model.fields.scalarList.map(field =>
-              printSet(`this.${field.name}`, `${printModelScalar(field)} ?? undefined`)
+              printSet(`this.${field.name}`, `${printModelScalar(field)}${field.nonNull ? "" : " ?? undefined"}`)
```
Comment (littlemooon) [inline packages/codegen-sdk/src/print-model.ts:87]: Quite a few places where we remove the `?`s

File: packages/codegen-sdk/src/print-model.ts [hunk 11]
```diff
@@ -76,37 +77,42 @@ function printModel(context: SdkPluginContext, model: SdkModel): string {
...
           printLines(
-            model.fields.object.map(field =>
-              printTernary(
-                printSet(`this.${field.name}`, `${Sdk.DATA_NAME}.${field.name}`),
-                `new ${field.object.name.value}(${Sdk.REQUEST_NAME}, ${Sdk.DATA_NAME}.${field.name}) `
-              )
-            )
+            model.fields.object.map(field => {
+              const operationCall = `new ${field.object.name.value}(${Sdk.REQUEST_NAME}, ${Sdk.DATA_NAME}.${field.name})`;
+              return field.nonNull
```
Comment (littlemooon) [inline packages/codegen-sdk/src/print-model.ts:94]: Also remove ternaries where nonNull

File: packages/codegen-sdk/src/print-scalar.ts [hunk 12]
```diff
@@ -42,9 +42,9 @@ export function printModelScalar(field: SdkModelField): string {
 
   switch (field.type) {
     case Sdk.SCALAR_DATE_TYPE:
-      return `${parseDateFunction}(${fieldName})`;
+      return field.nonNull ? `${parseDateFunction}(${fieldName}) ?? new Date()` : `${parseDateFunction}(${fieldName})`;
```
Comment (littlemooon) [inline packages/codegen-sdk/src/print-scalar.ts:45]: Default the scalar parsing functions when parsing non null fields

File: packages/sdk/src/_tests/readme.test.ts [hunk 13]
```diff
@@ -57,12 +57,12 @@ describe("readme.md", () => {
         /** CODE_SECTION:1_1:START */
         async function getMyIssues() {
           const me = await linearClient.viewer;
-          const myIssues = await me?.assignedIssues();
+          const myIssues = await me.assignedIssues();
 
-          if (myIssues?.nodes?.length) {
-            myIssues?.nodes?.map(issue => console.log(`${me?.displayName} has issue: ${issue?.title}`));
```
Comment (littlemooon) [inline packages/sdk/src/_tests/readme.test.ts:63]: Tests and readme no longer need optional chaining in most cases

**Review by eldh — APPROVED** [review]

**Pull request merged**